### ISSUE TITLE
memory leak and speed concerns

### ISSUE THREAD

**Issue by pseudotensor** [h2oai/datatable#1596]
```
import numpy as np
import lightgbm_gpu as lgb
import scipy
import pandas as pd
from sklearn.utils import shuffle
from h2oaicore.metrics import def_rmse
import datatable as dt

def set_dt_col(train_dt, name, value):
    if isinstance(name, int):
        name = train_dt.names[name]
    train_dt[:, name] = dt.Frame(value)
    return train_dt

nrow = 4000
ncol = 5000
X = np.random.randn(nrow, ncol)
y = np.random.randn(nrow)
model = lgb.LGBMRegressor(objective='regression', n_jobs=20)  # 40 very slow
model.fit(X, y)

X_dt = dt.Frame(X)
cols_actual = list(X_dt.names)

do_numpy = False
score_f = def_rmse
preds = model.predict(X)
main_metric = score_f(actual=y, predicted=preds)
seed = 1234
def go():
    feature_importances = {}
    for n in range(ncol):
        print(n, flush=True)
        if do_numpy:
            shuf = shuffle(X[:,n].ravel())
            X_tmp = X # .copy()
            X_tmp[:,n] = shuf
            new_preds = model.predict(X_tmp)
            metric = score_f(actual=y, predicted=new_preds)
            col = "C" + str(n)
            feature_importances[col] = main_metric - metric
        else:
            col = cols_actual[n]
            shuf = shuffle(X_dt[:, col].to_numpy().ravel(), random_state=seed)
            X_tmp = set_dt_col(dt.Frame(X_dt), col, shuf)
            new_preds = model.predict(X_tmp)

            metric = score_f(actual=y, predicted=new_preds)
            feature_importances[col] = main_metric - metric
    return feature_importances

print(go())
```

Related to permutation variable importance.

If do_numpy = False, so it uses dt, then I see the resident memory slowly creep up from about 0.8GB to 1.6GB at n=1800 etc.  By n=4000 it's using 2.7GB.

If I use do_numpy = True, so it uses no dt, then I see resident memory never change over all n.

I thought at one point I only saw with LightGBM and not xgboost, but I'm not sure.

Unit tests like this numpy version by Microsoft show LightGBM not itself leaking: 
https://github.com/Microsoft/LightGBM/issues/1968

These 2 cases aren't doing exactly the same thing in that the numpy version keeps shuffling the same original X, while the dt version I think has essentially 2 copies, but the other original X_dt columns are not modified.  But @st-pasha you can confirm.

One can add the X_tmp = X.copy(), but it's not quite fair.  It makes a full copy, while dt should get away with only overwriting a single column.

Perhaps the flaw is how we are using dt and the frames?

**Comment by pseudotensor** [comment]
The other concern is that the dt version is much slower than the numpy version.

**Comment by pseudotensor** [comment]
Also, note one can comment out the metric and feature_importances stuff, it doesn't matter.

**Comment by pseudotensor** [comment]
`time python creep_alt.py` with do_numpy=True takes 2m48s
`time python creep_alt.py` with do_numpy=False takes 2m48s


Note that DAI relies heavily upon similar operations. @arnocandel , so this can be an issue overall.

**Comment by arnocandel** [comment]
`set_dt_col(dt.Frame(X_dt), col, shuf)` is just dt.Frame(X_dt)[col] = shuf, and returning that, so maybe Frame copy during `dt.Frame(X_dt)` is the issue?

**Comment by arnocandel** [comment]
memory increase is probably just lazy gc, maybe try with more rows and closer to system mem limit

**Comment by arnocandel** [comment]
We also need to replace the shuffle by doing native dt stuff like making index array and indirecting dt[shuf_idx, :] to shuffle

**Comment by pseudotensor** [comment]
Well, it could be some interaction, I'm not sure.  AFAIK, a simple lightgbm unit test shows no leak.  But I can try to see if still leaks without the predict.

**Comment by st-pasha** [comment]
In terms of speed, my hypothesis is that the following happens:

`datatable` is a column-oriented store, with every column's data located at its own memory address. This allows us to swap, slice, select, add or remove columns effortlessly. However when you pass a datatable Frame to another library, it tries to convert it into a numpy array, which is contiguous. At that point datatable has to copy all its columns' data into a single contiguous buffer to satisfy numpy's requirements.
At the same time, `numpy` cannot swap the columns back-and-forth the way `datatable` does: when a column gets replaced, the new data has to be written on top of the existing one. But in the end they don't have to gather the data into a single contiguous chunk: the data is already contiguous.

**Comment by pseudotensor** [comment]
Perhaps that's related to the leak.  In xgboost, we absorb the datatable directly, and I can't get that to leak.  But when passing directly to lightgbm (or in DAI when we convert datatable to numpy ourselves), then there is a leak.

**Comment by pseudotensor** [comment]
@st-pasha the extra dt.Frame() calls are harmless speed-wise?  I recall in past we checked and they were not harmless.

**Comment by st-pasha** [comment]
`dt.Frame()` does shallow copying, so even for a 5000-column frame it takes only 6ms (so roughly 1µs per column).
Conversion to numpy involves the so-called "buffer" interface in python. The ownership of the object is temporarily transferred to the caller, and they are responsible for calling the `buffer_release()` in the end. This a very leak-prone process, as you can imagine.

**Comment by pseudotensor** [comment]
Ya, no creeping memory usage if I use do_numpy=False but do:
```
-          new_preds = model.predict(X_tmp)
+           X_lgbm = X_tmp.to_numpy()
+          new_preds = model.predict(X_lgbm)
```

**Comment by st-pasha** [comment]
That's good news, it means the leak is almost surely on LightGBM's side.

**Comment by pseudotensor** [comment]
Well, I'm not confident the buffer facility is not itself leaky, this thing, right?

https://jakevdp.github.io/blog/2014/05/05/introduction-to-the-python-buffer-protocol/

**Comment by pseudotensor** [comment]
```
In both cases, PyBuffer_Release() must be called when the buffer isn’t needed anymore. Failure to do so could lead to various issues such as resource leaks.
```
@st-pasha Ok, I think you must be right.  LightGBM must not be releasing the buffers when they read-in the dt object.  The buffer doesn't go out of scope, so they would need to specifically release it (maybe not in C land, but at least in python land) once they are done with the buffer-sourced object.  RIght?

**Comment by guolinke** [comment]
Actually, the datatabel object is not converted to 2d numpy object, but CSR format in LightGBM side.
refer to
https://github.com/Microsoft/LightGBM/blob/b37065dbd5510079aac341a34895f4854c782fcb/python-package/lightgbm/basic.py#L469

@pseudotensor 

```
-           X_lgbm = X_tmp.to_numpy()
+          X_lgbm  = scipy.sparse.csr_matrix(X_tmp)
+          new_preds = model.predict(X_lgbm)
```

maybe you can try this.

**Comment by pseudotensor** [comment]
I could, although I assume you anyways take the input and optimally store the data like XGBoost with its dmatrix, right? (with regard to missings)

**Comment by guolinke** [comment]
@pseudotensor actually not. In prediction, we just pass the raw data over trained tree one time. There is no store operation.

**Comment by arnocandel** [comment]
this leaks:
```
import datatable as dt
X=dt.fread("weather.csv")
while (True):
	X_g = X[:, {n: dt.sum(dt.f[n]) for i, n in enumerate(X[:, int].names)}]
	X_g.key = X_g.names
	X = X[:, :, dt.join(X_g)]
```

**Comment by arnocandel** [comment]
after a minute or so:
`18214 arno      20   0 5334060 2.172g  11528 R  4338  0.9 137:33.50 python                `

**Comment by arnocandel** [comment]
same here:
```
import datatable as dt
X=dt.fread("weather.csv")
while (True):
	X_g = X[:, {n: dt.sum(dt.f[n]) for i, n in enumerate(X[:, int].names)}]
	X_g.key = X_g.names
	Y = X[:, :, dt.join(X_g)]
	del X_g
	del Y
```

**Comment by st-pasha** [comment]
Yes, I can reproduce this on my Mac too. The amount leaked seems to be rather small, about 100 bytes per iteration:
```
>>> for i in range(10000):
...     X = X[:, :, dt.join(X_g)]
...     if i % 1000 == 0:
...         print(memory_usage())
... 
16.53515625
16.8359375
16.94140625
17.0546875
17.16015625
17.26953125
17.375
17.484375
17.59375
17.703125
```
It doesn't seem to depend on the size of either `X` or `X_g`. I'll try to figure out where this comes from.
